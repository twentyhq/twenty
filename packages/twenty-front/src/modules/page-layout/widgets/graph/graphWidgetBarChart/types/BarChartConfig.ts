import { type GraphColorScheme } from '@/page-layout/widgets/graph/types/GraphColorScheme';

export type BarChartConfig = {
  key: string;
  indexValue: string | number;
  gradientId: string;
  colorScheme: GraphColorScheme;
};
