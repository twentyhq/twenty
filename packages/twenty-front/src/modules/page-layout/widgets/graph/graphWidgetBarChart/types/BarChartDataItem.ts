import { type BarDatum } from '@nivo/bar';

export type BarChartDataItem = BarDatum & {
  to?: string;
};
