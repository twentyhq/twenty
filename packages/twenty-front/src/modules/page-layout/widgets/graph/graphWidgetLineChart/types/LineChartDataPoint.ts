export type LineChartDataPoint = {
  x: number | string | Date;
  y: number | null;
  to?: string;
};
