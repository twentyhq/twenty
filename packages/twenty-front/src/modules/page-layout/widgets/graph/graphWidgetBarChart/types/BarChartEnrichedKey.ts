import { type GraphColorScheme } from '@/page-layout/widgets/graph/types/GraphColorScheme';

export type BarChartEnrichedKey = {
  key: string;
  colorScheme: GraphColorScheme;
  label: string;
};
