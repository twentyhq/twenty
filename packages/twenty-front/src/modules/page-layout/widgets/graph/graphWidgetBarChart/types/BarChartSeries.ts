import { type GraphColor } from '@/page-layout/widgets/graph/types/GraphColor';

export type BarChartSeries = {
  key: string;
  label?: string;
  color?: GraphColor;
};
