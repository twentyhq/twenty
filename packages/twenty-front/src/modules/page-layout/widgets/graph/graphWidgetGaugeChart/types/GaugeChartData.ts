import { type GraphColor } from '@/page-layout/widgets/graph/types/GraphColor';

export type GaugeChartData = {
  value: number;
  min: number;
  max: number;
  color?: GraphColor;
  to?: string;
  label?: string;
};
