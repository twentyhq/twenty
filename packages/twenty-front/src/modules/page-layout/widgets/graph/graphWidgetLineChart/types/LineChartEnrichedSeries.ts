import { type LineChartSeries } from '@/page-layout/widgets/graph/graphWidgetLineChart/types/LineChartSeries';
import { type GraphColorScheme } from '@/page-layout/widgets/graph/types/GraphColorScheme';

export type LineChartEnrichedSeries = LineChartSeries & {
  colorScheme: GraphColorScheme;
  gradientId: string;
  shouldEnableArea: boolean;
  label: string;
};
