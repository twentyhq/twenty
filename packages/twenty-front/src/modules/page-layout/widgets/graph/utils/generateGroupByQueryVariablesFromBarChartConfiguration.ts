import { type ObjectMetadataItem } from '@/object-metadata/types/ObjectMetadataItem';
import { getGroupByOrderBy } from '@/page-layout/widgets/graph/utils/getGroupByOrderBy';
import {
  type AggregateOrderByWithGroupByField,
  type ObjectRecordOrderByForCompositeField,
  type ObjectRecordOrderByForScalarField,
  type ObjectRecordOrderByWithGroupByDateField,
} from 'twenty-shared/types';
import { isDefined } from 'twenty-shared/utils';
import { type BarChartConfiguration } from '~/generated/graphql';
import { buildGroupByFieldObject } from './buildGroupByFieldObject';

export const generateGroupByQueryVariablesFromBarChartConfiguration = ({
  objectMetadataItem,
  barChartConfiguration,
  aggregateOperation,
}: {
  objectMetadataItem: ObjectMetadataItem;
  barChartConfiguration: BarChartConfiguration;
  aggregateOperation?: string;
}) => {
  const groupByFieldXId =
    barChartConfiguration.primaryAxisGroupByFieldMetadataId;

  const groupByFieldYId =
    barChartConfiguration.secondaryAxisGroupByFieldMetadataId;

  const groupBySubFieldNameX =
    barChartConfiguration.primaryAxisGroupBySubFieldName ?? undefined;

  const groupBySubFieldNameY =
    barChartConfiguration.secondaryAxisGroupBySubFieldName ?? undefined;

  const groupByFieldX = objectMetadataItem.fields.find(
    (field) => field.id === groupByFieldXId,
  );

  const groupByFieldY = isDefined(groupByFieldYId)
    ? objectMetadataItem.fields.find((field) => field.id === groupByFieldYId)
    : undefined;

  if (!isDefined(groupByFieldX) || !isDefined(groupByFieldXId)) {
    throw new Error(
      `Field with id ${groupByFieldXId} not found in object metadata`,
    );
  }

  const groupBy: Array<
    Record<string, boolean | Record<string, boolean | string>>
  > = [];

  groupBy.push(
    buildGroupByFieldObject({
      field: groupByFieldX,
      subFieldName: groupBySubFieldNameX,
      dateGranularity:
        barChartConfiguration.primaryAxisDateGranularity ?? undefined,
    }),
  );

  if (isDefined(groupByFieldY)) {
    groupBy.push(
      buildGroupByFieldObject({
        field: groupByFieldY,
        subFieldName: groupBySubFieldNameY,
        dateGranularity:
          barChartConfiguration.secondaryAxisDateGranularity ?? undefined,
      }),
    );
  }

  const orderBy: Array<
    | AggregateOrderByWithGroupByField
    | ObjectRecordOrderByForScalarField
    | ObjectRecordOrderByWithGroupByDateField
    | ObjectRecordOrderByForCompositeField
  > = [];

  if (isDefined(barChartConfiguration.primaryAxisOrderBy)) {
    orderBy.push(
      getGroupByOrderBy({
        graphOrderBy: barChartConfiguration.primaryAxisOrderBy,
        groupByField: groupByFieldX,
        groupBySubFieldName:
          barChartConfiguration.primaryAxisGroupBySubFieldName,
        aggregateOperation,
        dateGranularity:
          barChartConfiguration.primaryAxisDateGranularity ?? undefined,
      }),
    );
  }
  if (
    isDefined(groupByFieldY) &&
    isDefined(barChartConfiguration.secondaryAxisOrderBy)
  ) {
    orderBy.push(
      getGroupByOrderBy({
        graphOrderBy: barChartConfiguration.secondaryAxisOrderBy,
        groupByField: groupByFieldY,
        groupBySubFieldName:
          barChartConfiguration.secondaryAxisGroupBySubFieldName,
        aggregateOperation,
        dateGranularity:
          barChartConfiguration.secondaryAxisDateGranularity ?? undefined,
      }),
    );
  }

  return {
    groupBy,
    ...(orderBy.length > 0 && { orderBy }),
  };
};
