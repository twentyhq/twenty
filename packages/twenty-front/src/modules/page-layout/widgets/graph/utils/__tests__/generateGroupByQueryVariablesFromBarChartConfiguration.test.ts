import { type ObjectMetadataItem } from '@/object-metadata/types/ObjectMetadataItem';
import {
  type BarChartConfiguration,
  ExtendedAggregateOperations,
  FieldMetadataType,
  GraphType,
} from '~/generated-metadata/graphql';
import { generateGroupByQueryVariablesFromBarChartConfiguration } from '../generateGroupByQueryVariablesFromBarChartConfiguration';

describe('generateGroupByQueryVariablesFromBarChartConfiguration', () => {
  const mockObjectMetadataItem: ObjectMetadataItem = {
    id: 'obj-1',
    nameSingular: 'opportunity',
    namePlural: 'opportunities',
    fields: [
      {
        id: 'field-1',
        name: 'stage',
        type: FieldMetadataType.TEXT,
      },
      {
        id: 'field-2',
        name: 'owner',
        type: FieldMetadataType.RELATION,
      },
      {
        id: 'field-3',
        name: 'createdAt',
        type: FieldMetadataType.DATE_TIME,
      },
      {
        id: 'field-4',
        name: 'name',
        type: FieldMetadataType.FULL_NAME,
      },
    ],
  } as ObjectMetadataItem;

  const buildConfiguration = (
    overrides: Partial<BarChartConfiguration>,
  ): BarChartConfiguration =>
    ({
      __typename: 'BarChartConfiguration',
      aggregateFieldMetadataId: 'aggregate-field',
      aggregateOperation: ExtendedAggregateOperations.COUNT,
      graphType: GraphType.VERTICAL_BAR,
      primaryAxisGroup: 'field-1',
      ...overrides,
    }) as BarChartConfiguration;

  describe('Vertical bar configuration', () => {
    it('should generate variables with single groupBy field', () => {
      const result = generateGroupByQueryVariablesFromBarChartConfiguration({
        objectMetadataItem: mockObjectMetadataItem,
        barChartConfiguration: buildConfiguration({
          graphType: GraphType.VERTICAL_BAR,
          primaryAxisGroup: 'field-1',
          primaryAxisSubFieldName: null,
        }),
      });

      expect(result).toMatchSnapshot();
    });

    it('should generate variables with two groupBy fields', () => {
      const result = generateGroupByQueryVariablesFromBarChartConfiguration({
        objectMetadataItem: mockObjectMetadataItem,
        barChartConfiguration: buildConfiguration({
          graphType: GraphType.VERTICAL_BAR,
          primaryAxisGroup: 'field-1',
          primaryAxisSubFieldName: null,
          secondaryAxisGroup: 'field-2',
          secondaryAxisSubFieldName: null,
        }),
      });

      expect(result).toMatchSnapshot();
    });

    it('should generate variables with composite field', () => {
      const result = generateGroupByQueryVariablesFromBarChartConfiguration({
        objectMetadataItem: mockObjectMetadataItem,
        barChartConfiguration: buildConfiguration({
          graphType: GraphType.VERTICAL_BAR,
          primaryAxisGroup: 'field-4',
          primaryAxisSubFieldName: 'firstName',
        }),
      });

      expect(result).toMatchSnapshot();
    });
  });

  describe('Horizontal bar configuration', () => {
    it('should generate variables with single groupBy field', () => {
      const result = generateGroupByQueryVariablesFromBarChartConfiguration({
        objectMetadataItem: mockObjectMetadataItem,
        barChartConfiguration: buildConfiguration({
          graphType: GraphType.HORIZONTAL_BAR,
          primaryAxisGroup: 'field-1',
          primaryAxisSubFieldName: null,
        }),
      });

      expect(result).toMatchSnapshot();
    });

    it('should generate variables with secondary axis', () => {
      const result = generateGroupByQueryVariablesFromBarChartConfiguration({
        objectMetadataItem: mockObjectMetadataItem,
        barChartConfiguration: buildConfiguration({
          graphType: GraphType.HORIZONTAL_BAR,
          primaryAxisGroup: 'field-1',
          secondaryAxisGroup: 'field-2',
        }),
      });

      expect(result).toMatchSnapshot();
    });
  });

  it('should throw error when primary axis field not found', () => {
    expect(() =>
      generateGroupByQueryVariablesFromBarChartConfiguration({
        objectMetadataItem: mockObjectMetadataItem,
        barChartConfiguration: buildConfiguration({
          primaryAxisGroup: 'invalid-field',
        }),
      }),
    ).toThrow('Field with id invalid-field not found in object metadata');
  });
});
