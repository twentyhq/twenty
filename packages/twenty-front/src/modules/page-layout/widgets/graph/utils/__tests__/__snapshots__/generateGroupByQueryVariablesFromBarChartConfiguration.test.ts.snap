// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateGroupByQueryVariablesFromBarChartConfiguration Horizontal bar configuration should generate variables with secondary axis 1`] = `
{
  "groupBy": [
    {
      "stage": true,
    },
    {
      "ownerId": true,
    },
  ],
}
`;

exports[`generateGroupByQueryVariablesFromBarChartConfiguration Horizontal bar configuration should generate variables with single groupBy field 1`] = `
{
  "groupBy": [
    {
      "stage": true,
    },
  ],
}
`;

exports[`generateGroupByQueryVariablesFromBarChartConfiguration Vertical bar configuration should generate variables with composite field 1`] = `
{
  "groupBy": [
    {
      "name": {
        "firstName": true,
      },
    },
  ],
}
`;

exports[`generateGroupByQueryVariablesFromBarChartConfiguration Vertical bar configuration should generate variables with single groupBy field 1`] = `
{
  "groupBy": [
    {
      "stage": true,
    },
  ],
}
`;

exports[`generateGroupByQueryVariablesFromBarChartConfiguration Vertical bar configuration should generate variables with two groupBy fields 1`] = `
{
  "groupBy": [
    {
      "stage": true,
    },
    {
      "ownerId": true,
    },
  ],
}
`;
